cppcache/src/TcrMessage.cpp

```diff
@@ -480,14 +480,14 @@ void TcrMessage::readLongPart(DataInput& input, uint64_t* intValue) {
 
 const std::string TcrMessage::readStringPart(DataInput& input) {
   int32_t stringLength = input.readInt32();
-  auto stringBuffer = std::unique_ptr<char[]>(new char[stringLength + 1]);
-  stringBuffer.get()[stringLength] = '\0';
+  std::vector<char> stringBuffer(stringLength + 1);
+  stringBuffer[stringLength] = '\0';
   if (input.read()) {
     throw Exception("String is not an object");
   }
-  input.readBytesOnly(reinterpret_cast<int8_t*>(stringBuffer.get()),
+  input.readBytesOnly(reinterpret_cast<int8_t*>(stringBuffer.data()),
```

Comment:
This isn't your problem but maybe you want to tackle it.
I dug a little deeper in here. It felt wrong that we were just reading the raw bytes into the string. As it turns out my gut was correct. The server either sends an ASCII only string (by setting of global flag that defaults false) or Java Modified UTF-8 (see `Message.addStringPart`). Since Java Modified UTF-8 isn't a standard we have special code on the C++ client to transcode it to UTF-8. This block of code should really be invoking a call to read Java Modified UTF-8. It would be great if we could just invoke it right on `DataInput` but the method that can read JMUTF-8 is private and expects to read the length too but when encode in `Part` the length is in the part header not the body. I think the solution is to just read the JMUTF directly using `JavaModifiedUtf8::decode`, which gives you the `std::u16string` (easy transcode), then transcode that to `std::string` in UTF8 using `to_utf8` in util/string.hpp.

Reply:
In case it isn't clear, reading the JMUTF-8 will safely take care of the ASCII cases since ASCII is a 1:1 encoding into JMUTF-8 and UTF-8.

Reply:
I tried to fix the bug following your directions.
I did not write a test because I did not fully understand the problem so I'd appreciate some help to write a proper test.

Reply:
I doubt there is an existing test that covers this. Given that we are talking to the server here we would need an integration test. I think if you create a test that creates a cq with non-ASCII characters and you get some update events for the CQ, then you should pass through this code. Maybe you can modify an existing CQ test and port it to the new integration framework.

Reply:
There are CQ tests both in the old and new frameworks that when executed, pass through the updated code.
I modified both tests to use CQs with names that contained non-ASCII characters and ran them with and without the changes in readStringPart but saw no difference in the results.
It looks as if the client behaves in the same way with the changes.
Am I missing something?

Reply:
Doh! My bad. just being non-ASCII isn’t enough. The fix is correct and the old code would have worked too For most values. The server is sending JMUTF-8 which looks a lot like UTF-8 except around NUL and characters outside the basic plane. The name would have to have a NUL (u0x00) character Or very high Unicode character in it before the old code would have produced invalid UTF-8. I’ll poke at one of the tests to see if I can make it fail and get back to you.

Reply:
Ok, I can reproduce the failure with string literal `u8"SimpleCQ\xF0\x90\x90\x80"` as the CQ name. This UTF-8 sequence produces the U+10400 character "𐐀", which has two UTF-16 code points and results in 6 bytes of JMUTF-8 rather then 4 bytes of UTF-8.

I also found a race condition in the new CqTest but I will write that up separately.

Reply:
Great! I was not able to reproduce it with a chinese character.